### PR TITLE
Restrict generic key type to only those supported

### PR DESCRIPTION
**Pull request by pete-woods**
Without this, you can specify key types that aren't supported and then have to deal with panics, rather than more pleasant compile-time warnings.

### CODE REVIEW

**Review by mangalaman93 — APPROVED** [review]

**Comment by mangalaman93** [comment]
Thanks for the PR. The only problem that I see with this is that any new type addition requires changes to the code and upgrading ristretto.

**Comment by pete-woods** [comment]
> Thanks for the PR. The only problem that I see with this is that any new type addition requires changes to the code and upgrading ristretto.

I believe this is already the case (see linked code): https://github.com/dgraph-io/ristretto/blob/main/z/z.go#L39

File: z/z.go [hunk 1]
```diff
@@ -23,18 +23,19 @@ import (
...
 // we already know that appending bytes isn't reliable for generating a
 // second hash (see Ristretto PR #88).
 // We also know that while the Go runtime has a runtime memhash128
 // function, it's not possible to use it to generate [2]uint64 or
 // anything resembling a 128bit hash, even though that's exactly what
 // we need in this situation.
-func KeyToHash[K any](key K) (uint64, uint64) {
+func KeyToHash[K Key](key K) (uint64, uint64) {
 	keyAsAny := any(key)
-	if keyAsAny == nil {
```
Comment (pete-woods) [inline z/z.go:35]: I *think* the linter correctly identified this check as redundant now (it forced me to remove it).

File: z/z.go [hunk 2]
```diff
@@ -23,18 +23,19 @@ import (
...
 // TODO: Figure out a way to re-use memhash for the second uint64 hash,
 // we already know that appending bytes isn't reliable for generating a
 // second hash (see Ristretto PR #88).
 // We also know that while the Go runtime has a runtime memhash128
 // function, it's not possible to use it to generate [2]uint64 or
 // anything resembling a 128bit hash, even though that's exactly what
 // we need in this situation.
-func KeyToHash[K any](key K) (uint64, uint64) {
+func KeyToHash[K Key](key K) (uint64, uint64) {
 	keyAsAny := any(key)
```
Comment (mangalaman93) [inline z/z.go:38]: you probably don't need to do this anymore

Comment (pete-woods) [reply z/z.go:38]: Unfortunately, that's against the Go language spec (see generics proposal): https://go.googlesource.com/proposal/+/refs/heads/master/design/43651-type-parameters.md#why-not-permit-type-assertions-on-values-whose-type-is-a-type-parameter

TL;DR, you'll get an error `cannot use type switch on type parameter value` when trying to perform a type switch on a parameterized type.

**Review by mangalaman93 — APPROVED** [review]

**Comment by pete-woods** [comment]
Thanks for your time! 🙇
